Common/Securities/CashAccountMarginModel.cs

```diff
@@ -0,0 +1,242 @@
+            var targetCurrency = string.Empty;
+            var openOrdersTotalValue = 0m;
+            var openOrdersTotalQuantity = 0m;
+            var symbolDirectionPairs = new Dictionary<Symbol, OrderDirection>();
+
+            if (baseCurrency != null)
+            {
+                // find the target currency for the requested direction and the securities potentially involved
+                targetCurrency = direction == OrderDirection.Buy
+                    ? security.QuoteCurrency.Symbol
+                    : baseCurrency.BaseCurrencySymbol;
+
+                foreach (var portfolioSecurity in portfolio.Securities.Values)
+                {
+                    var basePortfolioSecurity = portfolioSecurity as IBaseCurrencySymbol;
+                    if (basePortfolioSecurity == null) continue;
+
+                    if (basePortfolioSecurity.BaseCurrencySymbol == targetCurrency)
+                    {
+                        symbolDirectionPairs.Add(portfolioSecurity.Symbol, OrderDirection.Sell);
+                    }
+                    else if (portfolioSecurity.QuoteCurrency.Symbol == targetCurrency)
+                    {
+                        symbolDirectionPairs.Add(portfolioSecurity.Symbol, OrderDirection.Buy);
+                    }
+                }
+            }
+
+            // fetch open orders with matching symbol/side
+            var openOrders = portfolio.Transactions.GetOpenOrders(x =>
+                {
+                    OrderDirection dir;
+                    return symbolDirectionPairs.Count == 0 ||
+                           symbolDirectionPairs.TryGetValue(x.Symbol, out dir) && dir == x.Direction;
+                }
+            );
+
+            // calculate total value for selected orders
+            foreach (var order in openOrders)
+            {
+                var orderSecurity = portfolio.Securities[order.Symbol];
+                var orderBaseCurrency = orderSecurity as IBaseCurrencySymbol;
+
+                if (baseCurrency != null && orderBaseCurrency != null)
+                {
+                    var orderPrice = 0m;
+                    switch (order.Type)
+                    {
+                        case OrderType.Limit:
+                            orderPrice = ((LimitOrder) order).LimitPrice;
+                            break;
+                        case OrderType.StopMarket:
+                            orderPrice = ((StopMarketOrder) order).StopPrice;
+                            break;
+                        case OrderType.StopLimit:
+                            orderPrice = ((StopLimitOrder) order).LimitPrice;
+                            break;
+                    }
+
+                    // convert order value to target currency
+                    decimal orderValue;
+                    if (orderSecurity.QuoteCurrency.Symbol == targetCurrency)
+                    {
+                        orderValue = orderPrice * order.Quantity;
```

Comment:
Why aren't we using `order.GetValue(security)` here?  Where possible, I like to use polymorphism rather than conditional branching (ifs/switches) -- polymorphism encapsulates the logic where and conditional branching leaks to logic. An blurb from [refactoring guru](https://refactoring.guru/replace-conditional-with-polymorphism) on refactoring switches into polymorphic calls and benefits/ect.